Rewrite Supporting IPv6-only networks section

oskar456 opened the pull request:
This simplifies the section about IPv6-only networks while hopefully addressing comments from #91:

 - Remove mentions of IPv6-mostly, since the host where Happy Eyeballs runs will never be IPv6-mostly, it would be either Dual-stack or IPv6-only.
 - Use the presence of PREF64 as the only detection mechanism of NAT64
 - Note that RFC 7050 should be used, but only when RFC 8781 failed
 - Prescribe handling of IPv4 addresses by first looking up the routing table (needed for VPNs) and then synthesizing IPv6. This already covers the former text on IPv4 literals, broken AAAA records and VPN case
 - Add section on dealing with DNS64: basically detect and undo since it is not needed for HE.
 - Removed the `Last Resort Local Synthesis Delay` configurable variable. If we want HE to function properly without DNS64, every IPv4-only destination looks exactly like "Broken AAAA record" case and we definitely don't want 2 second delay every single time.

oskar456 committed:
Rewrite Supporting IPv6-only networks section

This simplifies the section about IPv6-only networks while hopefully
addressing comments from #91:

 - Remove mentions of IPv6-mostly, since the host where Happy Eyeballs
   runs will never be IPv6-mostly, it would be either Dual-stack or
   IPv6-only.
 - Use the presence of PREF64 as the only detection mechanism of NAT64
 - Note that RFC 7050 should be used, but only when RFC 8781 failed
 - Prescribe handling of IPv4 addresses by first looking up the routing
   table (needed for VPNs) and then synthesizing IPv6. This already covers
   the former text on IPv4 literals, broken AAAA records and VPN case
 - Add section on dealing with DNS64: basically detect and undo since it
   is not needed for HE.
 - Removed the `Last Resort Local Synthesis Delay` configurable variable
   If we want HE to function properly without DNS64, every
   IPv4-only destination looks exactly like "Broken AAAA record" case
   and we definitely don't want 2 sencond delay every single time.

philsbln reviewed (commented):
I like the rewrite, but it still needs refining. 

We let's have the editors take a look at this one and #93 and have them decide which one would be the better starting point. Either one works for me.

oskar456 committed:
Apply suggestions from @philsbln

Co-authored-by: Philipp S. Tiesel <philipp@tiesel.net>

tfpauly commented on draft-ietf-happy-happyeyeballs-v3.md:
```diff
@@ -576,135 +576,77 @@ If an address is added to the list, it should be sorted into the list
+the following:
+
+  1. Consult IPv4 routing table. If there is a valid route towards requested
+     destination, send native IPv4 data.
```
I agree with @oskar456 that trying to detect CLAT is a complexity we don't necessarily want to add as a step in the spec. It is possible depending on the support in the routing lookup. Maybe we can mention that we can filter out CLAT if possible?

furry13 replied:
>send native IPv4 data

Is any data being sent at this stage? I'd assume it's more about "use that IPv4 address as it is"?

oskar456 replied:
> > send native IPv4 data
> 
> Is any data being sent at this stage? I'd assume it's more about "use that IPv4 address as it is"?

Good point!

oskar456 replied:
> Rethinking it, detecting CLAT defies the purpose of being able to fall back to IPv4 upstream.

I don't think so. Falling back to IPv4 upstream happens in that case in step number 2 - that is by synthesising IPv6 address with NAT64 prefix and destination IPv4 address. It is only the matter of efficiency. If CLAT route cannot be avoided, step number one will always succeed (since CLAT would provide a default route) and step 2 would be never reached. But that is suboptimal. Ignoring CLAT routes in step 1 and generating native IPv6 towards NAT64 in step 2 is more efficient.

oskar456 committed:
Apply suggestions from @furry13 and @tfpauly

 - Put back IPv6-Mostly into the heading
 - Explain that IPv6-Mostly is either dual-stack or IPv6-only
 - Better introductionary text about DNS64 and 464XLAT
 - Use generic "NAT64 prefix" in place of "PREF64" which is specific for
   RFC 8781
 - Link RFC 9872 regarding NAT64 prefix

oskar456 commented on draft-ietf-happy-happyeyeballs-v3.md:
```diff
@@ -576,135 +576,75 @@ If an address is added to the list, it should be sorted into the list
-to detect one of these networks is to check if the network offers
-routable IPv6 addressing, does not offer routable IPv4 addressing, and
-offers a DNS resolver address.
+## Discovering and Utilizing PREF64 {#pref64-detection}
```
Applied, thanks!

philsbln commented on draft-ietf-happy-happyeyeballs-v3.md:
```diff
@@ -576,135 +576,77 @@ If an address is added to the list, it should be sorted into the list
-synthesize an IPv6 address, as if the resolved IPv4 address were
-provided by the application ({{literals}}).
+During the first step, the Happy Eyeballs MAY choose to ignore the IPv4 default
+route if such a route leads to the CLAT translator. The IPv6 addresses created
```
I still thing that sentence should be removed. As state above, this defies the purpose of being able to fall back to IPv4 upstream.

oskar456 replied:
I am actually thinking of changing this to SHOULD:

> During the first step, the Happy Eyeballs SHOULD ignore the IPv4 default route if such a route leads to the CLAT translator and such detection is feasible.

philsbln reviewed (approved):
I think it is ready, but still think the sentence abut ignoring CLAT should be removed.

oskar456 committed:
Fixup "send native IPv4 data"

philsbln reviewed (changes requested):
After Discussions on Montreal, let's hold this until we ran this by a design team. 

My hope is that it will simplify dramatically if we just treat known NAT64 as IPv4. 
If you don't look too closely, CLAT is already IPv4.
Whether to prefer self-synthesized NAT64 addresses over CLAT and native-IPv4 over CLAT seems to be a minor optimization to me which probably 99% of the use-cases can ignore.
